rs/artifact_pool/src/ingress_pool/peer_counter.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use ic_interfaces::ingress_pool::IngressPoolObject;
+use ic_logger::{warn, ReplicaLogger};
+use ic_types::{CountBytes, NodeId};
+
+#[derive(Clone)]
+/// Keeps track of the number of messages and total size of the messages per peer.
+pub(super) struct PeerCounters {
+    bytes_counters: PeerCounter,
+    count_counters: PeerCounter,
+}
+
+impl PeerCounters {
+    /// Create a new [`PeerCounters`] without any limits on the number/size of messages.
+    pub(super) fn new(log: ReplicaLogger) -> Self {
+        Self {
+            bytes_counters: PeerCounter::new(log.clone()),
+            count_counters: PeerCounter::new(log),
+        }
+    }
+
+    /// Create a new [`PeerCounters`] with the provided limits on the number/size of messages.
+    pub(super) fn new_with_limits(log: ReplicaLogger, max_bytes: usize, max_count: usize) -> Self {
+        Self {
+            bytes_counters: PeerCounter::new_with_limit(log.clone(), max_bytes),
+            count_counters: PeerCounter::new_with_limit(log, max_count),
+        }
+    }
+
+    pub(super) fn observe(&mut self, ingress_message: &IngressPoolObject) {
+        self.count_counters.add(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .add(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    pub(super) fn forget(&mut self, ingress_message: &IngressPoolObject) {
+        self.count_counters.subtract(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .subtract(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    /// Checks whether either the number of messages or their total size exceed the
+    /// respective limits.
+    pub(super) fn exceeds_limit(&self, peer_id: &NodeId) -> bool {
+        self.bytes_counters.exceeds_limit(peer_id) || self.count_counters.exceeds_limit(peer_id)
+    }
+
+    pub(super) fn count_total_bytes(&self) -> usize {
+        self.bytes_counters.total()
+    }
+}
+
+#[derive(Clone)]
+struct PeerCounter {
+    counter_per_peer: BTreeMap<NodeId, usize>,
```

Comment:
Any reason for `BTreeMap`?

Reply:
No reasons, do you think `HashMap` would work better?

Reply:
I was wondering if I am missing sth (some ordering requirement). No need to change it.

Reply:
Ok